File: at-docs-website/translate_docs.py
import os
import openai
import concurrent.futures
from tqdm import tqdm

# 调用 OpenAI API 翻译内容
def translate_content(content, target_language="en", api_key=None, api_url=None, model="gpt-4"):
    """调用 OpenAI API 翻译内容"""
    try:
        openai.api_key = api_key
        if api_url:
            openai.api_base = api_url
        
        response = openai.ChatCompletion.create(
            model=model,
            messages=[
                {"role": "system", "content": f"You are a professional translator. Translate this content into {target_language}."},
                {"role": "user", "content": content}
            ]
        )
        return response['choices'][0]['message']['content']
    except Exception as e:
        return None, str(e)

def process_file(file_path, output_dir, target_language, api_key, api_url, model):
    """处理单个文件的翻译"""
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            content = f.read()

        translated_content = translate_content(content, target_language, api_key, api_url, model)
        if translated_content is None:
            return False, file_path

        output_path = os.path.join(output_dir, os.path.basename(file_path))
        with open(output_path, "w", encoding="utf-8") as f:
            f.write(translated_content)

        return True, file_path
    except Exception as e:
        return False, file_path

def translate_markdown_files(input_dir, output_dir, target_language="en", max_workers=3, api_key=None, api_url=None, model="gpt-4"):
    """翻译目录下的所有 Markdown 文件"""
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    files = [os.path.join(root, file) for root, _, files in os.walk(input_dir) for file in files if file.endswith(".md")]
    total_files = len(files)

    progress = tqdm(total=total_files, desc="Translating", unit="file")

    success_count = 0
    fail_count = 0
    failed_files = []

    def translate_worker(file_path):
        success, file = process_file(file_path, output_dir, target_language, api_key, api_url, model)
        return success, file

    with concurrent.futures.ThreadPoolExecutor(max_workers=max_workers) as executor:
        futures = {executor.submit(translate_worker, file): file for file in files}
        for future in concurrent.futures.as_completed(futures):
            success, file = future.result()
            if success:
                success_count += 1
            else:
                fail_count += 1
                failed_files.append(file)
            progress.update(1)

    progress.close()

    print(f"\n翻译完成！")
    print(f"总文件数: {total_files}")
    print(f"成功翻译: {success_count}")
    print(f"失败翻译: {fail_count}")
    if failed_files:
        print("以下文件翻译失败:")
        for file in failed_files:
            print(f"- {file}")

if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser(description="批量翻译 Markdown 文件")
    parser.add_argument("input_dir", help="源语言文件目录")
    parser.add_argument("output_dir", help="目标语言文件目录")
    parser.add_argument("target_language", help="目标语言代码，如 'en'、'zh' 等")
    parser.add_argument("--concurrency", type=int, default=3, help="并发翻译文件数，默认 3")
    parser.add_argument("--api_key", required=True, help="OpenAI API 密钥")
    parser.add_argument("--api_url", default=None, help="OpenAI API URL，自定义时使用")
    parser.add_argument("--model", default="gpt-4", help="使用的 OpenAI 模型名称，默认 'gpt-4'")

    args = parser.parse_args()

    translate_markdown_files(
        input_dir=args.input_dir,
        output_dir=args.output_dir,
        target_language=args.target_language,
        max_workers=args.concurrency,
        api_key=args.api_key,
        api_url=args.api_url,
        model=args.model
    )

File: at-docs-website/translate_docs.js
const fs = require('fs');
const path = require('path');
const { Configuration, OpenAIApi } = require('openai');
const async = require('async');
const cliProgress = require('cli-progress');

async function translateContent(content, targetLanguage, apiKey, apiUrl, model) {
    try {
        const configuration = new Configuration({
            apiKey: apiKey,
            basePath: apiUrl || 'https://api.openai.com/v1',
        });
        const openai = new OpenAIApi(configuration);

        const response = await openai.createChatCompletion({
            model: model,
            messages: [
                { role: 'system', content: `You are a professional translator. Translate this content into ${targetLanguage}.` },
                { role: 'user', content: content },
            ],
        });

        return response.data.choices[0].message.content;
    } catch (error) {
        return { error: error.message };
    }
}

async function processFile(filePath, outputDir, targetLanguage, apiKey, apiUrl, model) {
    try {
        const content = fs.readFileSync(filePath, 'utf-8');
        const translatedContent = await translateContent(content, targetLanguage, apiKey, apiUrl, model);

        if (translatedContent.error) {
            return { success: false, file: filePath, error: translatedContent.error };
        }

        const outputFilePath = path.join(outputDir, path.basename(filePath));
        fs.writeFileSync(outputFilePath, translatedContent, 'utf-8');

        return { success: true, file: filePath };
    } catch (error) {
        return { success: false, file: filePath, error: error.message };
    }
}

async function translateMarkdownFiles(inputDir, outputDir, targetLanguage, concurrency, apiKey, apiUrl, model) {
    if (!fs.existsSync(outputDir)) {
        fs.mkdirSync(outputDir, { recursive: true });
    }

    const files = [];
    function getFilesRecursive(dir) {
        const entries = fs.readdirSync(dir, { withFileTypes: true });
        for (const entry of entries) {
            const fullPath = path.join(dir, entry.name);
            if (entry.isDirectory()) {
                getFilesRecursive(fullPath);
            } else if (entry.isFile() && (entry.name.endsWith('.md') || entry.name.endsWith('.mdx'))) {
                files.push(fullPath);
            }
        }
    }
    
    getFilesRecursive(inputDir);

    const totalFiles = files.length;
    const bar = new cliProgress.SingleBar({}, cliProgress.Presets.shades_classic);
    bar.start(totalFiles, 0);

    let successCount = 0;
    let failCount = 0;
    const failedFiles = [];

    await async.eachLimit(files, concurrency, async (filePath) => {
        const result = await processFile(filePath, outputDir, targetLanguage, apiKey, apiUrl, model);
        if (result.success) {
            successCount++;
        } else {
            failCount++;
            failedFiles.push({ file: result.file, error: result.error });
        }
        bar.increment();
    });

    bar.stop();

    console.log(`\n翻译完成！`);
    console.log(`总文件数: ${totalFiles}`);
    console.log(`成功翻译: ${successCount}`);
    console.log(`失败翻译: ${failCount}`);

    if (failedFiles.length > 0) {
        console.log('以下文件翻译失败:');
        failedFiles.forEach(({ file, error }) => {
            console.log(`- ${file}: ${error}`);
        });
    }
}

// 命令行解析
const yargs = require('yargs/yargs');
const { hideBin } = require('yargs/helpers');

const argv = yargs(hideBin(process.argv))
    .option('inputDir', { type: 'string', demandOption: true, describe: '源语言文件目录' })
    .option('outputDir', { type: 'string', demandOption: true, describe: '目标语言文件目录' })
    .option('targetLanguage', { type: 'string', demandOption: true, describe: '目标语言代码，如 "en" 或 "zh"' })
    .option('concurrency', { type: 'number', default: 3, describe: '并发翻译文件数，默认 3' })
    .option('apiKey', { type: 'string', demandOption: true, describe: 'OpenAI API 密钥' })
    .option('apiUrl', { type: 'string', describe: 'OpenAI API URL，自定义时使用' })
    .option('model', { type: 'string', default: 'gpt-4', describe: '使用的 OpenAI 模型名称，默认 "gpt-4"' })
    .help()
    .argv;

// 执行翻译
translateMarkdownFiles(
    argv.inputDir,
    argv.outputDir,
    argv.targetLanguage,
    argv.concurrency,
    argv.apiKey,
    argv.apiUrl,
    argv.model
);
